File: 1-virtual-waiting-room-and-queuing/1-java/src/main/java/com/starci/waitingroom/WaitingRoomController.java
package com.starci.waitingroom;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
public class WaitingRoomController {

    private final WaitingRoomService service;

    public WaitingRoomController(WaitingRoomService service) {
        this.service = service;
    }

    @GetMapping("/health")
    public Map<String, String> health() {
        return Map.of("status", "ok");
    }

    @GetMapping("/api/waitingroom/token")
    public WaitingRoomService.TokenResult token() {
        return service.enqueue();
    }

    @GetMapping("/api/waitingroom/position")
    public WaitingRoomService.PositionResult position(@RequestParam String token) {
        return service.position(token);
    }

    @PostMapping("/api/waitingroom/admit")
    @ResponseStatus(HttpStatus.OK)
    public WaitingRoomService.AdmitResult admit(@RequestBody(required = false) AdmitRequest body) {
        int count = (body == null || body.count() == null) ? 1 : Math.max(1, body.count());
        return service.admit(count);
    }

    @GetMapping("/api/waitingroom/status")
    public WaitingRoomService.StatusResult status(@RequestParam String token) {
        return service.status(token);
    }

    public record AdmitRequest(Integer count) {}
}

File: 1-virtual-waiting-room-and-queuing/0-typescript/src/waitingroom/waitingroom.controller.ts
import { Body, Controller, Get, HttpCode, Post, Query } from "@nestjs/common"
import { WaitingRoomService } from "./waitingroom.service"

interface AdmitBody {
    count?: number
}

@Controller("api/waitingroom")
export class WaitingRoomController {
    public constructor(private readonly service: WaitingRoomService) {}

    @Get("token")
    public async token(): Promise<{ token: string; position: number }> {
        return this.service.enqueue()
    }

    @Get("position")
    public async position(
        @Query("token") token: string,
    ): Promise<{ token: string; position: number | null; admitted: boolean }> {
        return this.service.position(token)
    }

    @Post("admit")
    @HttpCode(200)
    public async admit(
        @Body() body: AdmitBody,
    ): Promise<{ admitted: string[]; count: number }> {
        const count = Math.max(1, Math.floor(Number(body?.count ?? 1)))
        return this.service.admit(count)
    }

    @Get("status")
    public async status(
        @Query("token") token: string,
    ): Promise<{ token: string; state: string; position: number | null }> {
        return this.service.status(token)
    }
}
